extension/persistence/eclipselink/src/main/java/org/apache/polaris/extension/persistence/impl/eclipselink/PolarisEclipseLinkMetaStoreSessionImpl.java

```diff
@@ -257,6 +260,12 @@ public <T> T runInTransaction(
       } finally {
         localSession.remove();
       }
+    } catch (PersistenceException e) {
+      if (e.toString().toLowerCase(Locale.ROOT).contains("duplicate key")) {
+        throw new AlreadyExistsException(e, "Duplicate key error when persisting entity");
```

Comment:
Why do we want to throw an Iceberg exception here? This is polaris only right?

Reply:
The practical reason we use an Iceberg exception is because we already have exception mapping via `IcebergExceptionMapper` for this exception. This exception is currently used throughout Polaris ([e.g.](https://github.com/apache/polaris/blob/81ec424de5bf4e466660b15dca1778026be07e9e/polaris-service/src/main/java/org/apache/polaris/service/catalog/BasePolarisCatalog.java#L1542)) and so I think it's good to be consistent with that.

Reply:
Do we throw `jakarta.persistence.EntityExistsException` here or general PersistenceException?

Reply:
It's not a `EntityExistsException` unfortunately.

Reply:
I'd prefer adding the mapping to `IcebergExceptionMapper` and explicitly throwing `jakarta.persistence.EntityExistsException` here. Even though we have some loose leaking between jakarta exceptions vs Iceberg exceptions in some places it'll be nice to start segmenting them better depending on whether the code is agnostic to Iceberg.

Classes that implement Iceberg interfaces directly should stick to Iceberg exceptions, and ideally the metastore layer can be kept as Iceberg-agnostic as possible.

Reply:
IIRC from our offline discussion the dependency required to do that doesn't exist and we didn't want to add eclipse-link specific dependencies to the service or core modules. Rather, I would be supportive of each module potentially bringing its own exception mapper.

Of course, I think anything is better than a 500. The important thing is that we don't change the client-facing exception. We can always change the type here used to map to that exception

Reply:
Oh I see, didn't realize we hadn't already pulled in `jakarta.persistence-api`. Looks like we do pull it in as a `compileOnly` dependency in `polaris-core` and `polaris-service` already depends on `polaris-core` so it's kind of close.

Looking at the package it seems like the exceptions are indeed in `jakarta.persistence-api` so in theory it just makes the fatjar bigger if we do fully add it as a dependency in `polaris-service` (latest jar appears to just be ~200kB, so it's not *too* bad).

I guess I could go either way in this case.

Reply:
What do you think of letting modules bring their own exception mapper? I feel like that gives us the best of both words: good dependency hygiene and extensibility as new modules introduce new exceptions.

If you think that could work, I'll file an issue and then maybe merge this as-is. Otherwise, I can just add this dependency now too.

Reply:
I'd propose to have a Polaris exception mapping in the core module, so that the service module need only map Polaris exceptions to Iceberg exceptions. In that way, we decouple the extension from service module with a bridge from module core.

Reply:
> We will need to add a new Polaris exception in the core, which we may have to do anyways in the future.

I'm fine doing this, but we already use this AlreadyExistsException in many places. Would you have me refactor them all to use this new Polaris exception? Does that block this PR?

Reply:
> we already use this AlreadyExistsException in many places

We can fix that later. I'd still recommend to start create/use a Polaris exception in this PR so that no more new dependency is added to the extension module.

Reply:
Okay, I've rebased per the above. Going to close this thread, but let's re-open the discussion as needed on the new diff. Thanks!